components/leds.py

```diff
@@ -0,0 +1,289 @@
+    PULSE = auto()
+    FLASH = auto()
+    IDK = auto()
+    HALF_HALF = auto()
+
+
+class RobotState(Enum):
+    PICKED_UP_PIECE = auto()
+    LOOKING_FOR_PIECE = auto()
+    OTHER = auto()
+
+
+class Piece(Enum):
+    CONE = auto()
+    CUBE = auto()
+    NONE = auto()
+
+
+class PickupFromSide(Enum):
+    LEFT = auto()
+    RIGHT = auto()
+
+
+# creates a list of LEDData's from a List of (hsv col, repetitions)
+def make_pattern(
+    data: List[Tuple[LedColours, int]]
+) -> List[wpilib.AddressableLED.LEDData]:
+    pattern_data = []
+    for colour, number in data:
+        x = wpilib.AddressableLED.LEDData()
+        x.setHSV(*colour.value)
+        pattern_data += [x] * number
+    return pattern_data
+
+
+class StatusLights:
+    leds: wpilib.AddressableLED
+
+    FLASH_PERIOD = 0.4
+    PULSE_PERIOD = 1
+    PACMAN_PERIOD = 60
+    RAINBOW_PERIOD = 15
+    IDK_PERIOD = 60
+
+    def __init__(self):
+        self.led_length = 1
+
+        self.start_time = time.monotonic()
+
+        self.color = (0, 0, 0)
+
+        self.pattern = DisplayType.SOLID
+
+        self.side = PickupFromSide.RIGHT
+
+    def setup(self) -> None:
+        self.leds.setLength(self.led_length)
+        self.single_led_data = wpilib.AddressableLED.LEDData()
+        self.leds_data = [self.single_led_data] * self.led_length
+        self.leds.setData(self.leds_data)
+        self.leds.start()
+
+    def set_color(self, color: tuple[int, int, int] | LedColours):
+        self.color = color
+
+    def set_piece(self, piece: Piece):
+        match piece:
+            case Piece.CONE:
+                self.set_color(LedColours.YELLOW)
+            case Piece.CUBE:
+                self.set_color(LedColours.VIOLET)
+            case Piece.NONE:
+                self.set_color(LedColours.OFF)
+
+    def set_state(self, state: RobotState):
+        match state:
+            case RobotState.PICKED_UP_PIECE:
+                self.pattern = DisplayType.SOLID
+            case RobotState.LOOKING_FOR_PIECE:
+                self.pattern = DisplayType.HALF_HALF
+            case RobotState.OTHER:
+                self.set_color(LedColours.OFF)
+
+    def set_intake_side(self, side: PickupFromSide):
+        self.side = side
+
+    def set_display_pattern(self, pattern: DisplayType):
+        self.pattern = pattern
+
+    def set(
+        self,
+        color: tuple[int, int, int],
+        piece: Piece,
+        state: RobotState,
+        side: PickupFromSide,
+    ):
+        # self.set_color(color)
+        self.set_piece(piece)
+        self.set_state(state)
+        self.set_intake_side(side)
+
+    def calc_half(self):
+        led_data = []
+        for i in range(self.led_length):
+            if i < round(self.led_length / 2):
+                led_data.append(self.color)
+            else:
+                led_data.append(
+                    LedColours.RED
+                    if self.side == PickupFromSide.LEFT
+                    else LedColours.GREEN
```

Comment:
Have you considered attaching these colours to the `PickupFromSide` enum?

Reply:
It wouldn't much of a hassle, but this seems easier to read

Reply:
I'd argue that `self.side.colour` would be easier to read :shrug: